soroban-spec/src/gen/ts.rs

```diff
@@ -0,0 +1,339 @@
+    read::{from_wasm, FromWasmError},
+};
+
+pub mod boilerplate;
+pub mod wrapper;
+
+#[derive(thiserror::Error, Debug)]
+pub enum GenerateFromFileError {
+    #[error("reading file: {0}")]
+    Io(io::Error),
+    #[error("sha256 does not match, expected: {expected}")]
+    VerifySha256 { expected: String },
+    #[error("parsing contract spec: {0}")]
+    Parse(stellar_xdr::Error),
+    #[error("getting contract spec: {0}")]
+    GetSpec(FromWasmError),
+}
+
+pub fn generate_from_file(
+    file: &str,
+    verify_sha256: Option<&str>,
+) -> Result<String, GenerateFromFileError> {
+    // Read file.
+    let wasm = fs::read(file).map_err(GenerateFromFileError::Io)?;
+
+    // Produce hash for file.
+    let sha256 = Sha256::digest(&wasm);
+    let sha256 = format!("{sha256:x}");
+
+    if let Some(verify_sha256) = verify_sha256 {
+        if verify_sha256 != sha256 {
+            return Err(GenerateFromFileError::VerifySha256 { expected: sha256 });
+        }
+    }
+
+    // Generate code.
+    let json = generate_from_wasm(&wasm).map_err(GenerateFromFileError::GetSpec)?;
+    Ok(json)
+}
+
+pub fn generate_from_wasm(wasm: &[u8]) -> Result<String, FromWasmError> {
+    let spec = from_wasm(wasm)?;
+    let json = generate(&spec);
+    Ok(json)
+}
+
+pub fn generate(spec: &[ScSpecEntry]) -> String {
+    let collected: Vec<_> = spec.iter().map(Entry::from).collect();
+    collected.iter().map(entry_to_ts).join("\n")
+}
+
+fn doc_to_ts_doc(doc: &str) -> String {
+    if doc.is_empty() {
+        String::new()
+    } else {
+        let doc = doc.split('\n').join("\n * ");
+        format!(
+            r#"/**
+ * {doc}
+ */
+"#,
+        )
+    }
+}
+
+pub fn entry_to_ts(entry: &Entry) -> String {
+    match entry {
+        Entry::Function {
+            doc,
+            name,
+            inputs,
+            outputs,
+        } => {
+            let sign_me = doc
+                .contains("@signme")
+                .then_some("sign: true, ")
+                .unwrap_or_default();
+            let args = inputs
+                .iter()
+                .map(|i| format!("((i) => {})({})", type_to_js_xdr(&i.value), i.name))
+                .join(",\n        ");
+            let input = (!inputs.is_empty())
+                .then(|| {
+                    format!(
+                        "{{{}}}: {{{}}}",
+                        inputs.iter().map(func_input_to_arg_name).join(", "),
+                        inputs.iter().map(func_input_to_ts).join(", ")
+                    )
+                })
+                .unwrap_or_default();
+            let return_type = if outputs.is_empty() {
+                "".to_owned()
+            } else if outputs.len() == 1 {
+                format!(": Promise<{}>", type_to_ts(&outputs[0]))
+            } else {
+                format!(
+                    ": Promise<[{}]>>",
+                    outputs.iter().map(type_to_ts).join(", ")
+                )
```

Comment:
It's worth noting that `outputs` in `ScSpecFunctionV0` ([ref](https://github.com/stellar/rs-stellar-xdr/blob/a2f370c930bb94f7bc0c9ea0426ecef3700b90a9/src/next/generated.rs#L5751-L5756)) is a `VecM<ScSpecTypeDef, 1>`. The `1` means the maximum length is guaranteed to be `1`. So this else block can never be triggered. 

I guess this arises here because the way we're generating TS is by following a path of `ScSpecEntry` -> `json::Entry` -> `TS`, and in the conversion to `json::Entry` we lose information.

If the `ScSpecFunctionV0` was available here, you could do the following code to guarantee the maximum length of 1 and to get a compiler error if that ever changed:
```rust
let outputs: &VecM<ScSpecTypeDef, 1> = &sc_spec_function_v0.outputs;
```

Reply:
Initially to get something done I copied from the JSON stuff and liked the smaller surface area of the JSON type. However, as you point out it's a bad because it loses information. This to me suggests taking the time to generate to actual json schema so that these type constraints such as this one would be maintained. NEAR and CosmWasm use JSON schema, however with no agreed upon contract schema they all differ in their shape.

Issues like this are another reason that I still think producing a Wasm which can handle the JSON <-> XDR this also removes all dependencies from the generated code.

Reply:
The canonical form of contract interfaces is XDR, for better or worse, so I think if we have the interface in that form, which we do here, we should generate from the canonical form.

Independent of this we do need JSON <-> XDR, and we have it in the Rust XDR lib. I think there's value in seeing if we can exactly replicate the Rust XDR libs JSON <-> XDR in the js-stellar-base, but that's a big project and don't think we should make any of this work contingent on it.

Reply:
> we should generate from the canonical form

We can merge the PR as it is done now, but I think we should treat this as tech debt and circle back on it so there isn't that dependency / conversion path.